// @stryker-mutator/mocha-runner/src/MochaOptionsLoader.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var path = require("path");
var fs = require("fs");
var plugin_1 = require("@stryker-mutator/api/plugin");
var utils_1 = require("./utils");
var LibWrapper_1 = require("./LibWrapper");
/**
 * Subset of defaults for mocha options
 * @see https://github.com/mochajs/mocha/blob/master/lib/mocharc.json
 */
exports.DEFAULT_MOCHA_OPTIONS = Object.freeze({
    extension: ['js'],
    opts: './test/mocha.opts',
    spec: ['test'],
    timeout: 2000,
    ui: 'bdd'
});
var MochaOptionsLoader = /** @class */ (function () {
    function MochaOptionsLoader(log) {
        this.log = log;
    }
    MochaOptionsLoader.prototype.load = function (strykerOptions) {
        var mochaOptions = tslib_1.__assign({}, strykerOptions[utils_1.mochaOptionsKey]);
        return tslib_1.__assign({}, exports.DEFAULT_MOCHA_OPTIONS, this.loadMochaOptions(mochaOptions), mochaOptions);
    };
    MochaOptionsLoader.prototype.loadMochaOptions = function (overrides) {
        if (LibWrapper_1.default.loadOptions) {
            this.log.debug('Mocha >= 6 detected. Using mocha\'s `%s` to load mocha options', LibWrapper_1.default.loadOptions.name);
            return this.loadMocha6Options(overrides);
        }
        else {
            this.log.debug('Mocha < 6 detected. Using custom logic to parse mocha options');
            return this.loadLegacyMochaOptsFile(overrides.opts);
        }
    };
    MochaOptionsLoader.prototype.loadMocha6Options = function (overrides) {
        var args = utils_1.serializeArguments(overrides);
        var loadOptions = LibWrapper_1.default.loadOptions || (function () { return ({}); });
        var rawConfig = loadOptions(args) || {};
        if (this.log.isTraceEnabled()) {
            this.log.trace("Mocha: " + loadOptions.name + "([" + args.map(function (arg) { return "'" + arg + "'"; }).join(',') + "]) => " + JSON.stringify(rawConfig));
        }
        var options = utils_1.filterConfig(rawConfig);
        if (this.log.isDebugEnabled()) {
            this.log.debug("Loaded options: " + JSON.stringify(options, null, 2));
        }
        return options;
    };
    MochaOptionsLoader.prototype.loadLegacyMochaOptsFile = function (opts) {
        switch (typeof opts) {
            case 'boolean':
                this.log.debug('Not reading additional mochaOpts from a file');
                return {};
            case 'undefined':
                var defaultMochaOptsFileName = path.resolve(exports.DEFAULT_MOCHA_OPTIONS.opts);
                if (fs.existsSync(defaultMochaOptsFileName)) {
                    return this.readMochaOptsFile(defaultMochaOptsFileName);
                }
                else {
                    this.log.debug('No mocha opts file found, not loading additional mocha options (%s.opts was not defined).', utils_1.mochaOptionsKey);
                    return {};
                }
            case 'string':
                var optsFileName = path.resolve(opts);
                if (fs.existsSync(optsFileName)) {
                    return this.readMochaOptsFile(optsFileName);
                }
                else {
                    this.log.error("Could not load opts from \"" + optsFileName + "\". Please make sure opts file exists.");
                    return {};
                }
        }
    };
    MochaOptionsLoader.prototype.readMochaOptsFile = function (optsFileName) {
        this.log.info("Loading mochaOpts from \"" + optsFileName + "\"");
        return this.parseOptsFile(fs.readFileSync(optsFileName, 'utf8'));
    };
    MochaOptionsLoader.prototype.parseOptsFile = function (optsFileContent) {
        var _this = this;
        var options = optsFileContent.split('\n').map(function (val) { return val.trim(); });
        var mochaRunnerOptions = Object.create(null);
        options.forEach(function (option) {
            var _a;
            var args = option.split(' ').filter(Boolean);
            if (args[0]) {
                switch (args[0]) {
                    case '--require':
                    case '-r':
                        args.shift();
                        if (!mochaRunnerOptions.require) {
                            mochaRunnerOptions.require = [];
                        }
                        (_a = mochaRunnerOptions.require).push.apply(_a, args);
                        break;
                    case '--timeout':
                    case '-t':
                        mochaRunnerOptions.timeout = _this.parseNextInt(args);
                        break;
                    case '--async-only':
                    case '-A':
                        mochaRunnerOptions.asyncOnly = true;
                        break;
                    case '--ui':
                    case '-u':
                        mochaRunnerOptions.ui = _this.parseNextString(args);
                        break;
                    case '--grep':
                    case '-g':
                        var arg = "" + _this.parseNextString(args);
                        if (arg.startsWith('/') && arg.endsWith('/')) {
                            arg = arg.substring(1, arg.length - 1);
                        }
                        mochaRunnerOptions.grep = new RegExp(arg);
                        break;
                    default:
                        _this.log.debug("Ignoring option \"" + args[0] + "\" as it is not supported.");
                        break;
                }
            }
        });
        return mochaRunnerOptions;
    };
    MochaOptionsLoader.prototype.parseNextInt = function (args) {
        if (args.length > 1) {
            return parseInt(args[1], 10);
        }
        else {
            return undefined;
        }
    };
    MochaOptionsLoader.prototype.parseNextString = function (args) {
        if (args.length > 1) {
            return args[1];
        }
        else {
            return undefined;
        }
    };
    MochaOptionsLoader.inject = plugin_1.tokens(plugin_1.commonTokens.logger);
    return MochaOptionsLoader;
}());
exports.default = MochaOptionsLoader;
